File: examples/httpd/user/bak/anyterm/browser/anyterm.js
// browser/anyterm.js
// This file is part of Anyterm; see http://chezphil.org/anyterm/
// (C) 2005 Philip Endecott

// This is version $Name$
//   (if there is no version (e.g. V0-1) mentioned in the previous line,
//    this is probably a snapshot from between "official" releases.)

// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.


function min(a,b) {
  if (a<b) return a; else return b;
}

var term;
var kb_buf="";

var send_in_progress=0;
var pending_doc;

var poll_timer;
var min_poll_interval=1000;
var max_poll_interval=10000;
var poll_interval_escalator=1.2;
var poll_interval=min_poll_interval;


function increase_poll_interval() {
  poll_interval=min(poll_interval*poll_interval_escalator,
		    max_poll_interval);
}


function display(c) {
  // Display a character or sequence of characters on the terminal.
  // Will have to do proper terminal emulation, e.g. moving the cursor
  // eventually.  For now, totally naive implementation.
  term.value+=c;
}


function rcv() {
  // Called asynchronously when the pending_doc has returned from the server.
  send_in_progress=0;
  var root=pending_doc.firstChild;
  if (!root) {
    alert("Error: no data");
  }
  if (root.tagName=="error") {
    alert("Error: "+root.firstChild.data);
  } else if (root.tagName=="op") {
    if (root.firstChild) {
      var op = root.firstChild.data;
      display(op);
      poll_interval=min_poll_interval;
    } else {
      increase_poll_interval();
    }
    init_poll();
  } else {
    alert("Unrecognised response: "+root.tagName);
  }
}


function send() {
  send_in_progress=1;
  cancel_poll();
  pending_doc=Sarissa.getDomDocument();
  pending_doc.async = true;
  // pending_doc.method="POST";
  pending_doc.onreadystatechange = load_handler;
  pending_doc.load("foo.anyterm?k="+encodeURIComponent(kb_buf));
  if (kb_buf!="") {
    poll_interval=min_poll_interval;
  }
  kb_buf="";
}


function load_handler() {
  if (pending_doc.readyState==4) {
    rcv();
  }
}


function maybe_send() {
  if (send_in_progress) {
    return;
  }
  send();
}


function process_key(k) {
  kb_buf+=k;
  maybe_send();
}


function init_poll() {
  poll_timer=setTimeout(send,poll_interval);
}


function cancel_poll() {
  clearTimeout(poll_timer);
}


function keypressed(ev) {
  if (!ev) var ev=window.event;
  var kc;
  if (ev.keyCode) kc=ev.keyCode;
  if (ev.which)   kc=ev.which;
if (kc==13) kc=10;
  var k=String.fromCharCode(kc);
  //alert("Key "+k+" pressed");
  process_key(k);
  return false;
}


function initialise() {
  term=document.getElementById("term");
  term.onkeypress=keypressed;
  init_poll();
}


window.onload=initialise;

File: examples/httpd/user/bak/anyterm/apachemod/anyterm.c
// apachemod/anyterm.c
// This file is part of Anyterm; see http://chezphil.org/anyterm/
// (C) 2005 Philip Endecott

// This is version $Name$
//   (if there is no version (e.g. V0-1) mentioned in the previous line,
//    this is probably a snapshot from between "official" releases.)

// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.


#include "httpd.h"
#include "http_config.h"
#include "http_core.h"
#include "http_log.h"
#include "http_main.h"
#include "http_protocol.h"
#include "http_request.h"
#include "util_script.h"
#include "http_connection.h"
#include "apr_strings.h"

#include <pty.h> 
#include <unistd.h>



module AP_MODULE_DECLARE_DATA anyterm;


static int master_fd;


static void report_error(request_rec* r, char* what)
{
  ap_rputs("<error>",r);
  ap_rputs(strerror(errno),r);
  ap_rputs(" when ",r);
  ap_rputs(what,r);
  ap_rputs("</error>",r);
}


static void anyterm_open(request_rec* r)
{
  pid_t pid=forkpty(&master_fd, NULL, NULL, NULL);
  if (pid==-1) {
    report_error(r,"forkpty()");
    return;
  }
  if (pid==0) {
    //execl("/bin/login","login",NULL);
    execl("/bin/bash","bash",NULL);
    report_error(r,"execl() returned");
    return;
  }
  int rc=fcntl(master_fd,F_SETFL,O_NONBLOCK);
  if (rc==-1) {
    report_error(r,"fcntl(F_SETFL,O_NONBLOCK)");
    return;
  }
  ap_rputs("<open/>",r);
}

static void anyterm_close(request_rec* r)
{
  int rc=close(master_fd);
  if (rc==-1) {
    report_error(r,"close()");
    return;
  }
  ap_rputs("<close/>",r);
}



static void anyterm_input(request_rec* r, const char* k)
{
  int l=strlen(k);
  if (l>0) {
    char* dk=malloc(l+1);
    if (!dk) {
      report_error(r,"malloc()");
    }
    const char* p=k;
    char* q=dk;
    char hexbuf[3];
    while(*p) {
      if (*p=='+') {
	*q=' ';
	p++;
      } else if (*p=='%') {
	hexbuf[0]=p[1];
	hexbuf[1]=p[2];
	hexbuf[2]='\0';
	*q=strtol(hexbuf,NULL,16);
	p+=3;
      } else {
	*q=*p;
	p++;
      }
      q++;
    }
    int dl=q-dk;
    int rc=write(master_fd,dk,dl);
    free(dk);
    if (rc!=dl) {
      report_error(r,"write()");
      return;
    }
  }
  int first=1;
  static char buf[256];
  int n;
  while(1) {
    n=read(master_fd,buf,sizeof(buf)-1);
    if (n==-1) {
      if (errno==EAGAIN) {
	n=0;
      } else {
	report_error(r,"read()");
	if (first) {
	  return;
	} else {
	  break;
	}
      }
    }
    if (first) {
      ap_rputs("<op>",r);
      first=0;
    }
    if (n==0) {
      break;
    }
    buf[n]='\0';
    ap_rputs(buf,r);
  }
  ap_rputs("</op>",r);
}


static int anyterm_handler(request_rec* r)
{
  if (strcmp(r->handler,"anyterm")) {
    return DECLINED;
  }

  ap_set_content_type(r,"text/xml");
  if (r->header_only) {
    return OK;
  }

  char* args;
  if (r->args) {
    args=r->args;
  } else {
    args="";
  }

  ap_rputs("<?xml version=\"1.0\"?>\n",r);

  if (strcmp(args,"open")==0) {
    anyterm_open(r);
  } else if (strcmp(args,"close")==0) {
    anyterm_close(r);
  } else if (strncmp(args,"k=",2)==0) {
    const char* k=args+2;
    anyterm_input(r,k);
  } else {
    ap_rputs("<error>invalid args '",r);
    ap_rputs(args,r);
    ap_rputs("'</error>",r);
  }
  
  return OK;
}


static void anyterm_register_hooks(apr_pool_t *p)
{
  ap_hook_handler(anyterm_handler, NULL, NULL, APR_HOOK_MIDDLE);
}



module AP_MODULE_DECLARE_DATA anyterm = {
  STANDARD20_MODULE_STUFF,
  NULL, /* per-directory config creator */
  NULL, /* dir config merger */
  NULL, /* server config creator */
  NULL, /* server config merger */
  NULL, /* command table */
  anyterm_register_hooks,  /* set up other request processing hooks */
};

